kore/src/Kore/Internal/Inj.hs

```diff
@@ -0,0 +1,143 @@
+{-|
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+
+-}
+
+module Kore.Internal.Inj
+    ( Inj (..)
+    , toSymbol, toApplication
+    -- * Exceptions
+    , UnorderedInj (..)
+    , unorderedInj
+    ) where
+
+import Control.DeepSeq
+import Control.Exception
+    ( Exception (..)
+    , throw
+    )
+import Data.Function
+    ( (&)
+    )
+import qualified Data.Function as Function
+import Data.Hashable
+import qualified Data.Text.Prettyprint.Doc as Pretty
+import Data.Typeable
+    ( Typeable
+    )
+import qualified Generics.SOP as SOP
+import qualified GHC.Generics as GHC
+
+import Debug
+import Kore.Attribute.Pattern.FreeVariables
+    ( FreeVariables
+    )
+import qualified Kore.Attribute.Symbol as Attribute
+import Kore.Attribute.Synthetic
+    ( Synthetic (..)
+    )
+import Kore.Internal.ApplicationSorts
+import Kore.Internal.Symbol
+import Kore.Sort
+import Kore.Syntax.Application
+    ( Application (..)
+    )
+import Kore.Unparser
+
+data Inj a =
+    Inj
+        { injConstructor :: !Id
+        , injFrom        :: !Sort
+        , injTo          :: !Sort
+        , injAttributes  :: !Attribute.Symbol
+        , injChild       :: !a
+        }
+    deriving (GHC.Generic, Show)
+    deriving (Functor, Foldable, Traversable)
+
+instance Eq a => Eq (Inj a) where
```

Comment:
Below I'll assume that you're not comparing attributes in order to make things faster (if not, then what's the actual reason?).

Is it reasonable to exclude all attributes from equality testing? The synthetic ones and the ones determined by the symbol definition should be fine, should we add a comment to `Attribute.Symbol` that we should pay attention to that? I.e. that we should not add one which "mutable" in a way similar to the `simplified` bit?

As a general solution, maybe we should also compare attributes, but we should have a special comparison function which ignores the part that's uniquely determined (i.e. for `Attribute.Symbol` it should always return that the operands are equal). Whenever we change the attribute, we would also be forced to change the comparison function, so we would make a decision about that change.

Reply:
> As a general solution, maybe we should also compare attributes, but we should have a special comparison function which ignores the part that's uniquely determined

I considered having the `Eq` instance `assert` that the attributes are the same if the symbol names are the same. I only didn't do that because the `Eq Symbol` instance doesn't do that.